saltstack/salt
# PoC: the Liquid Salt system

We have discussed some time ago under #41879 regarding the possibility of a pluggable system to allow Salt to load (further) configuration options for various systems.

This system can have the advantage of managing the configuration opts (or some of them) from a central point, without having to regenerate the config files on hundreds/thousands of Minions. Of course, this comes with the risk that with a single click you can break your entire infrastructure. But it's the same risk when using the `ext_pillar`, so the user assumes the risk (we'll probably need to document that).

That can be a very sensitive subject, as @whiteinge mentioned:

> Managing the configs from a GUI or automating deploying config changes is a bit tricker -- since the daemon will need to be restarted to pick up any changes we need to be certain in advance that the daemon will restart successfully. Suggestions have been to perform a pre-check validation step (#4352) or to fail gracefully. But if there is a problem in the generated config such as bad syntax or incorrect parameters it can cause the master to not start or to traceback.

What I have so far is very far away from achieving these goals, for the moment I only want to present what I currently have, and discuss if it makes sense to you to continue on this path.

The base idea is that you can put at least (or only) the following lines in your Master / Minion config file and the rest of the options are fetched from the systems you want, e.g.,:

```/etc/salt/master```
```yaml
liquid:
  - yaml:
      path: https://example.com/yaml-api
  - mongo:
      host: example.com
      port: 49017
      db: common_opts
  - mongo:
      host: example.com
      port: 49017
      db: master_only_opts
```

I have called this system _Liquid_ (because https://en.wikipedia.org/wiki/Molten_salt_reactor#Liquid_salt_very_high_temperature_reactor), but maybe _Molten_ could be a better name? (Molten-salt fueling options). Let me know if we should rename. :-)

### Design choices

- You still need to provide the `/etc/salt/master` and `/etc/salt/minion` files, that have at least the liquid config.
- The opts collected via `liquid` will override anything else.
- The opts collected via `liquid` will be cached in a file called `_liquid.conf` under `/etc/salt/minion.d/_liquid.conf` (or `/etc/salt/master.d/_liquid.conf`). Is this a good enough place? I preferred so, because they are collected from there anyway, but it has the disadvantage that the user may turn off the liquid system and the opts will continue to be loaded from there. Probably we should save it elsewhere? If so, where?

#### What works so far:

- A pluggable system.
- Loads the opts from the sources configured.
- An execution module for CLI manipulations.

#### Challenges:

- During startup the configuration opts from the `/etc/salt/master` and `/etc/salt/minion` files are read several times (I added a error log just before this line: https://github.com/saltstack/salt/blob/develop/salt/config/__init__.py#L2140)

```
$ sudo salt-master -l error
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /root/.saltrc
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /root/.saltrc
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /root/.saltrc
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /root/.saltrc
[ERROR   ] Read config from /root/.saltrc
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
...
... snip ...
...
```

We probably should avoid hitting the remote services / APIs so much. However I added a `liquid_always_fetch` option that can be turned on if the user really wants that.

But I don't know how can I detect that the Minion/Master process is just starting up, and not loading the opts the 100-th time already? Is there any flag I missed? I added an artificial `__liquid_loaded` in the opts, but that turns to be a very weak way. Help please! :-)

If you like the idea and think there's value to continue this system, please let me know (and if possible help with the question(s) above), then I will finalize and document.

#### How to test this branch?

Just add the following in the Master or Minion config file:

```yaml
liquid:
  - yaml: {}
```

And you should see the following option loaded (e.g., using `test.get_opts`):

```yaml
some_test_data_from_the_liquid_yaml: True
```

Ping @cachedout @rallytime @garethgreenaway (from #41879)
